generator/src/main/java/com/scottlogic/deg/generator/constraints/atomic/AtomicConstraint.java

```diff
@@ -28,4 +28,8 @@ default AtomicConstraint accept(NodeVisitor visitor){
     }
 
     AtomicConstraint withRules(Set<RuleInformation> rules);
+
+    default boolean isSoftConstraint(){
```

Comment:
This would be repeated *everywhere*, seems like the wrong design considering each atomic constraint has exactly the same way of dealing with this? This isn't immediately extensible to (for example) greaterThan constraints without extra work. Could we have done this with a wrapper around atomic constraint instead similar to violatedAtomicConstraint?

Reply:
There are a number of different ways to implement this; including the decorator approach.

I went for this approach over the decorator for the following reasons:
- The decorator would need some ability to expose the underlying properties, which in a strongly typed language would require a decorator type for every underlying type?
   - e.g. a decorator for `IsStringLongerThanConstraint` which exposes `referenceValue` and a different decorator type for `IsOfTypeConstraint` which exposes `requiredType`.
- Having the `isSoftConstraint` 'property' on an `AtomicConstraint` provides the information that this is something `AtomicConstraint`s can be, as in this case
- Using a `default` interface implementation allows `AtomicConstraint`s to opt-in to this behaviour as they see fit, but not having to implement an empty/default method on every `AtomicConstraint`. So in response to your comment 'this would be repeated _everywhere_', I dont think this is true.

If there is a 'better' implementation then I'm all ears.

The alternatives could be:
- decoration, although as described above this could be cumbersome to implement
- a default/optional interface method (as implemented in this PR)
- a new interface which is optionally implemented, e.g. `SoftOrHardAtomicConstraint`, with the `boolean isSoftConstraint()` method defined on it
- another approach?